crates/libs/bindgen/src/rust/winrt_methods.rs

```diff
@@ -189,31 +232,53 @@ pub fn gen_upcall(
     inner: TokenStream,
     this: bool,
 ) -> TokenStream {
+    let noexcept = metadata::method_def_is_noexcept(sig.def);
+
     let invoke_args = sig
         .params
         .iter()
         .map(|param| gen_winrt_invoke_arg(writer, param));
+
     let this = if this {
         quote! { this, }
     } else {
         quote! {}
     };
 
     match &sig.return_type {
-        metadata::Type::Void => quote! {
-            #inner(#this #(#invoke_args,)*).into()
-        },
+        metadata::Type::Void => {
+            if noexcept {
+                quote! {
+                    #inner(#this #(#invoke_args,)*);
+                    windows_core::HRESULT(0)
+                }
+            } else {
+                quote! {
+                    #inner(#this #(#invoke_args,)*).into()
+                }
+            }
+        }
         _ if sig.return_type.is_winrt_array() => {
-            quote! {
-                match #inner(#this #(#invoke_args,)*) {
-                    Ok(ok__) => {
-                        let (ok_data__, ok_data_len__) = ok__.into_abi();
-                        // use `core::ptr::write` since `result` could be uninitialized
-                        core::ptr::write(result__, ok_data__);
-                        core::ptr::write(result_size__, ok_data_len__);
-                        windows_core::HRESULT(0)
+            if noexcept {
+                quote! {
+                    let ok__ = #inner(#this #(#invoke_args,)*);
+                    let (ok_data__, ok_data_len__) = ok__.into_abi();
+                    core::ptr::write(result__, ok_data__);
```

Comment:
This can also be `result__.write(ok_data__);`, for brevity.

Reply:
Will update this separately as it will produce a ton of code gen changes.